### PR TITLE
 Added parameters in splitObject()

### LINKED ISSUE #493: update splitObject and test cases for getList(), getListContent()
### Tasks to be done.

- [ ] Upgrade splitObject()
- [ ] Add tests for splitObject()
- [ ] Add tests for getList(), getListContent()

### PR DESCRIPTION
**Pull request by uniyalprashant9**
Fixes #493 

Added parameters in splitObject(path,filename)
1. path:- loaction of file.
2. filename: e.g:- "grocery.json".

Both parameters are string.

### CODE REVIEW

**Comment by uniyalprashant9** [comment]
Need to update readme.md after merge

**Comment by uniyalprashant9** [comment]
removed elements folder (we don't need it now) because the function which generated it is now changed.

**Comment by uniyalprashant9** [comment]
Will pull again after getting all modules optimized.

**Comment by uniyalprashant9** [comment]
now splitObject can split any json file into elements.
If object have name attrib then the splitted element will get the file name as the name attrib else will get a index as filename.

**Comment by atherdon** [comment]
if it's now can handle other cases as well - why we have line like this there: ``` if(typeof grocery.name === "undefined")```
grocery list can be only one case - we also can have pretty different datasets

**Comment by atherdon** [comment]
imagine that our file will be at separated place and you don't know what kind of data there, and your script with this function will be at another separated place. by separated places i mean for example repositories.

in any cases - it's a good task for mastering refactoring skills

**Comment by uniyalprashant9** [comment]
> if it's now can handle other cases as well - why we have line like this there: ` if(typeof grocery.name === "undefined")`
> grocery list can be only one case - we also can have pretty different datasets

if the name attribute is missing then this will give the splitted file name the index value in case if someone tries to split other json files which doesn't have a name attribute.

**Comment by uniyalprashant9** [comment]
> imagine that our file will be at separated place and you don't know what kind of data there, and your script with this function will be at another separated place. by separated places i mean for example repositories.
> 
> in any cases - it's a good task for mastering refactoring skills

I already made all the modules independent. They just need the correct parameters and permission to work. :)

**Comment by atherdon** [comment]
> if the name attribute is missing then this will give the splitted file name the index value in case if someone tries to split other json files which doesn't have a name attribute.

that part - `grocery.name` was fine for case, when we splitting grocery lists. but for other cases - it's not cool. can you do it differently. spend some time and think how to do it

**Comment by atherdon** [comment]
Plus, can you also update readme file - so people will understand that we can use split object for different cases and how to use it

**Comment by uniyalprashant9** [comment]
> > if the name attribute is missing then this will give the splitted file name the index value in case if someone tries to split other json files which doesn't have a name attribute.
> 
> that part - `grocery.name` was fine for case, when we splitting grocery lists. but for other cases - it's not cool. can you do it differently. spend some time and think how to do it

OK, I will keep it `grocery.name` and will update it later on.

**Comment by atherdon** [comment]
show me the code change - because now it's not clear how do you want to handle it

File: generator/writeFile.js [hunk 1]
```diff
@@ -59,27 +59,22 @@ function test () {
...
 
 function splitObject (path,file) {
-  var src = path; // Location of object file
-  var obj = file // Object file name
-  var srclen= src.length;
-  if(src.charAt(srclen-1)!=='/') src = src + '/'; //path correction
-  var location = src + obj // location of object file
-  let ddata = fs.readFileSync(location)
+  if(path.charAt(path.length-1)!=='/') path = path + '/'; //path correction
+  let ddata = fs.readFileSync(path + file)
```
Comment (atherdon) [inline generator/writeFile.js]: i don't like this variable name

File: generator/writeFile.js [hunk 2]
```diff
@@ -59,27 +59,22 @@ function test () {
...
   var len = grocery.length // Object Length
   var foldername = file.slice(0,-5) ;
-  foldername = foldername+ "_elements";
+  foldername = foldername + "_elements";
   var fileName;
-  if (!fs.existsSync(src + foldername)) fs.mkdirSync(src + foldername) // Check if Folder exists or not.
+  if (!fs.existsSync(path + foldername)) fs.mkdirSync(path + foldername) // Check if Folder exists or not.
   for (var i = 0; i < len; i++) {
     let temp = JSON.stringify(grocery[i]);
-    if(typeof grocery.name === "undefined") fileName = i + '.json'; // if no name attrib then index will be file name
```
Comment (atherdon) [inline generator/writeFile.js]: i actually like this if section
But what i don't like - that we have `grocery.name` inside of it

File: generator/writeFile.js [hunk 3]
```diff
@@ -59,27 +59,22 @@ function test () {
...
-  var obj = file // Object file name
-  var srclen= src.length;
-  if(src.charAt(srclen-1)!=='/') src = src + '/'; //path correction
-  var location = src + obj // location of object file
-  let ddata = fs.readFileSync(location)
+  if(path.charAt(path.length-1)!=='/') path = path + '/'; //path correction
+  let ddata = fs.readFileSync(path + file)
   let grocery = JSON.parse(ddata)
   var len = grocery.length // Object Length
   var foldername = file.slice(0,-5) ;
```
Comment (atherdon) [inline generator/writeFile.js]: replace with folderName

File: generator/writeFile.js [hunk 4]
```diff
@@ -59,27 +59,22 @@ function test () {
...
-  var src = path; // Location of object file
-  var obj = file // Object file name
-  var srclen= src.length;
-  if(src.charAt(srclen-1)!=='/') src = src + '/'; //path correction
-  var location = src + obj // location of object file
-  let ddata = fs.readFileSync(location)
+  if(path.charAt(path.length-1)!=='/') path = path + '/'; //path correction
+  let ddata = fs.readFileSync(path + file)
   let grocery = JSON.parse(ddata)
   var len = grocery.length // Object Length
```
Comment (atherdon) [inline generator/writeFile.js]: having `grocery` again

File: generator/writeFile.js [hunk 5]
```diff
@@ -59,27 +59,22 @@ function test () {
...
   var foldername = file.slice(0,-5) ;
-  foldername = foldername+ "_elements";
+  foldername = foldername + "_elements";
   var fileName;
-  if (!fs.existsSync(src + foldername)) fs.mkdirSync(src + foldername) // Check if Folder exists or not.
+  if (!fs.existsSync(path + foldername)) fs.mkdirSync(path + foldername) // Check if Folder exists or not.
```
Comment (atherdon) [inline generator/writeFile.js]: make it in few lines - it's not the best way for reading code

File: generator/writeFile.js [hunk 6]
```diff
@@ -59,27 +59,22 @@ function test () {
...
-    if(typeof grocery.name === "undefined") fileName = i + '.json'; // if no name attrib then index will be file name
-    else fileName = grocery[i].name + '.json';
+    fileName = grocery[i].name + '.json';
     fileName = fileName.replace(/ /g, '_'); // Replace space with underscore
     fileName = fileName.toLowerCase();
     // Checking if file already exists or not
-    if (!fs.existsSync(src + foldername + '/' + fileName)) fs.writeFileSync(src + foldername + '/' + fileName, temp);
-    else console.log('File <' + fileName + '> Already Exists.');
+    if (!fs.existsSync(path + foldername + '/' + fileName))
+      fs.writeFileSync(path + foldername + '/' + fileName, temp);
```
Comment (atherdon) [inline generator/writeFile.js]: you removed else statement

Comment (uniyalprashant9) [reply generator/writeFile.js]: adding else statement increased complexity :(

**Comment by atherdon** [comment]
then split this method into 2 simple methods

>

File: generator/writeFile.js (hunk 1 above)
Comment (uniyalprashant9) [reply generator/writeFile.js]: done

File: generator/writeFile.js (hunk 2 above)
Comment (uniyalprashant9) [reply generator/writeFile.js]: added it back

File: generator/writeFile.js (hunk 3 above)
Comment (uniyalprashant9) [reply generator/writeFile.js]: done

File: generator/writeFile.js (hunk 4 above)
Comment (uniyalprashant9) [reply generator/writeFile.js]: changed to fileData

File: generator/writeFile.js (hunk 5 above)
Comment (uniyalprashant9) [reply generator/writeFile.js]: done

File: generator/writeFile.js (hunk 6 above)
Comment (uniyalprashant9) [reply generator/writeFile.js]: i have used writeFile() that is already defined.

**Comment by uniyalprashant9** [comment]
@atherdon I made the required changes please review it.

**Comment by uniyalprashant9** [comment]
@atherdon  I made all the changes.
Screenshot: Testing with diffrent json files.
Screenshot from 2019-06-10 21-21-47

Screenshot from 2019-06-10 21-21-40

**Comment by uniyalprashant9** [comment]
Once changes are approved I will update readme.md

**Comment by uniyalprashant9** [comment]
readme.md updated.

**Comment by atherdon** [comment]
perfect. looks awesome! nice work

**Comment by uniyalprashant9** [comment]
@atherdon please give me another task.

**Comment by atherdon** [comment]
ok, this task will be also related to splitObject functionality, and then we'll switch into working on grocery lists. I also preparing a task for moving you into ML project.
Or if you want - we can have few other tasks at JS repository - just tell me.

use your function in order to split information(but not as each file == one object) for db-recipes.json and recipe-items-latest from https://github.com/GroceriStar/sd-plain/tree/master/dirty/Recipe

**Comment by uniyalprashant9** [comment]
@atherdon I will first complete #424 test part when we can move to  ML project.

**Comment by atherdon** [comment]
and we need to finish that split with 2 datasets that i shared with you

>

**Comment by uniyalprashant9** [comment]
> ok, this task will be also related to splitObject functionality, and then we'll switch into working on grocery lists. I also preparing a task for moving you into ML project.
> Or if you want - we can have few other tasks at JS repository - just tell me.
> 
> use your function in order to split information(but not as each file == one object) for db-recipes.json and recipe-items-latest from https://github.com/GroceriStar/sd-plain/tree/master/dirty/Recipe

I didn't understand `use your function in order to split information(but not as each file == one object)`, splitObject function splits the file into single elements only. Do we need to read the data from splitted file??

**Comment by uniyalprashant9** [comment]
> and we need to finish that split with 2 datasets that i shared with you
> […](#)

[https://github.com/GroceriStar/sd-plain/blob/master/dirty/Recipe/recipeitems-latest.json](url) this file is empty.

**Comment by atherdon** [comment]
>  splitObject function splits the file into single elements only.
can we extend that functionality? or maybe use your function as copy-paste for splitting big files as well?

> recipeitems-latest.json file is empty.
will take a look.

**Comment by uniyalprashant9** [comment]
> > splitObject function splits the file into single elements only.
> 
> can we extend that functionality? or maybe use your function as copy-paste for splitting big files as well?
> 

we have getListContent() in utils.js .... I can modify it so that it will work as a copy-paste.
Currently, it returns the content of all json files in a directory as a list.

**Comment by atherdon** [comment]
not sure if i'm understand your correctly, so i'll put everything in one comment.
1) your code works great no doubt to that. i ask you to made few changes because - a) i love to have code reusable, b) i hate when code used once and then someone will redo a same thing from scratch(it's a professional thing - i just a saw a lot of times when people doing the same thing over and over again from scratch - this is why i code last 2 years in JS - because it have a lot of modules done before)
and c) i also want to see how you'll deal with code review changes. because it's also big part of the process. and i think it'll also give you the ability to dive deeper in this project

2) as i want to be your code used more than just one time - i want to try to see that your code is actually capable to doing the same process(splitting data) but for our other datasets. this is why i asked for changes. for sure, if we have a whole database with 14k rows of data - we don't need 14k files at one folder. this is why i think we should modify/extend/upgrade your method so it will be able to cover all of our cases. 

3) so i'm a bit confused right now and you need to help me. Because i don't want to dive into it deeper and take a look on my own. i want to do it by your hands.

---

btw, maybe db-recipes are located at: https://github.com/GroceriStar/sd-plain/blob/master/dirty/Recipe/20170107-061401-recipeitems.json.gz i'm not quite sure right now - because i think that sd-plain was broken somehow. but we should try

**Comment by uniyalprashant9** [comment]
so should I add one more parameter that will tell if we need to split data or retrieve a particular row from the data?

**Comment by atherdon** [comment]
do the best way as you think it should be

>

**Comment by amogh-jrules** [comment]
> not sure if i'm understand your correctly, so i'll put everything in one comment.
> 
> 1. your code works great no doubt to that. i ask you to made few changes because - a) i love to have code reusable, b) i hate when code used once and then someone will redo a same thing from scratch(it's a professional thing - i just a saw a lot of times when people doing the same thing over and over again from scratch - this is why i code last 2 years in JS - because it have a lot of modules done before)
>    and c) i also want to see how you'll deal with code review changes. because it's also big part of the process. and i think it'll also give you the ability to dive deeper in this project
> 2. as i want to be your code used more than just one time - i want to try to see that your code is actually capable to doing the same process(splitting data) but for our other datasets. this is why i asked for changes. for sure, if we have a whole database with 14k rows of data - we don't need 14k files at one folder. this is why i think we should modify/extend/upgrade your method so it will be able to cover all of our cases.
> 3. so i'm a bit confused right now and you need to help me. Because i don't want to dive into it deeper and take a look on my own. i want to do it by your hands.
> 
> btw, maybe db-recipes are located at: https://github.com/GroceriStar/sd-plain/blob/master/dirty/Recipe/20170107-061401-recipeitems.json.gz i'm not quite sure right now - because i think that sd-plain was broken somehow. but we should try

Hey @uniyalprashant9 , I would like to contribute... mind if I help by completing @atherdon 's request?

**Comment by atherdon** [comment]
@amogh-jrules you can help by answering this question. as our file were splited into few files - can you tell me where we store them, because at quick search I cannot find them. And I also think this can be added to Readme - because someone else with has a similar task for sure will have problems to locate them

**Comment by atherdon** [comment]
> we have getListContent() in utils.js 

I actually don't know about it - @amogh-jrules can you add an explanation about this method at Readme.md here as well?

**Comment by uniyalprashant9** [comment]
> > we have getListContent() in utils.js
> 
> I actually don't know about it - @amogh-jrules can you add an explanation about this method at Readme.md here as well?

getList(path) --> it returns var(list) of all files in a directory.
getListContent(path) --> it returns a var (list) which have content of all json files int the directory.

I made these 2 functions to extend the capability of splitObject() or we can say that we can verify if splitObject() is working properly or not by these 2 modules.

I have already added about these in readme.md with splitObject().

**Comment by uniyalprashant9** [comment]
@atherdon I will merge getList() and getListContent() into a single module so that it will serve as both.

**Comment by atherdon** [comment]
@uniyalprashant9 what about files of grocery lists that we split? i cannot find them quickly

**Comment by uniyalprashant9** [comment]
@atherdon As we have updated splitObject those files needs to be updated.

I will add those files with adding a detailed information about splitObject and other modules.

**Comment by atherdon** [comment]
awesome

>

**Pull request merged**